### PR TITLE
Implement Prometheus endpoint for the parachain commander

### PR DESCRIPTION
**Pull request by vstakhov**
This PR implements Prometheus mode for the parachain commander tool, allowing to collect metrics about parachain(s) progression. So far, only one parachain is supported, however, that could easily be extended in future.

### CODE REVIEW

**Comment by sandreim** [comment]
Thanks @vstakhov ! This is really cool. I’ll test and review first thing when I get back in January.

**Comment by vstakhov** [comment]
Thank you, @sandreim! It's important to note that this pull request builds upon the previous two, so it may be more efficient to merge those first and then rebase this one for easier review. Please keep this in mind as you proceed.

**Comment by sandreim** [comment]
Gotcha! It would have been a bit easier to review if the PRs were stacked, but this also works.

**Comment by vstakhov** [comment]
It is stacked now, thanks @sandreim!

File: src/pc/prometheus.rs [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+			HistogramVec::new(
+				HistogramOpts::new("pc_block_time", "Block time for parachain measurements for relay parent blocks")
+					.buckets(HISTOGRAM_TIME_BUCKETS.into()),
+				&["parachain_id"],
+			)?,
+			registry,
+		)?,
+		slow_avail_count: prometheus_endpoint::register(
+			IntCounterVec::new(
+				Opts::new("pc_slow_available_count", "Number of slow availability events"),
```
Comment (sandreim) [inline src/pc/prometheus.rs:227]:
```suggestion
				Opts::new("pc_slow_available_count", "Number of slow availability events. We consider it slow when the relay chain block bitfield entries amounts to less than 2/3 one bits for the availability core to which the parachain is assigned"),
```

File: src/pc/prometheus.rs [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+				&["parachain_id"],
+			)?,
+			registry,
+		)?,
+		low_bitfields_count: prometheus_endpoint::register(
+			IntCounterVec::new(
+				Opts::new("pc_low_bitfields_count", "Number of low bitfields events"),
```
Comment (sandreim) [inline src/pc/prometheus.rs:234]:
```suggestion
				Opts::new("pc_low_bitfields_count", "Number of low bitfields count events. This happens when a block author received the signed bitfields from less than 2/3 of the para validators"),
```

File: src/pc/prometheus.rs [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+			)?,
+			registry,
+		)?,
+		bitfields: prometheus_endpoint::register(
+			IntGaugeVec::new(Opts::new("pc_bitfields_count", "Number of bitfields"), &["parachain_id"]).unwrap(),
+			registry,
+		)?,
+		included_times: prometheus_endpoint::register(
+			HistogramVec::new(
+				HistogramOpts::new("pc_included_time", "Average included time in relay parent blocks")
```
Comment (sandreim) [inline src/pc/prometheus.rs:245]:
```suggestion
				HistogramOpts::new("pc_included_time", "Parachain block time measured in relay chain blocks.")
```

File: src/pc/mod.rs [hunk 4]
```diff
@@ -59,13 +76,17 @@ pub(crate) struct ParachainCommanderOptions {
 	/// Defines subscription mode
 	#[clap(short = 's', long = "subscribe-mode", default_value_t, value_enum)]
 	pub subscribe_mode: SubxtSubscriptionMode,
+	/// Mode of running - CLI/Prometheus.
+	#[clap(subcommand)]
+	mode: Option<ParachainCommanderMode>,
```
[the remark this answers is not in the document]
Comment (vstakhov) [reply src/pc/mod.rs:81]: As far as I understand, that's the way to make a subcommand optional (e.g. to preserve backwards compatibility).

[the remark this answers is not in the document]
Comment (sandreim) [reply src/pc/mod.rs:81]: A comment would be enough in this case

File: src/pc/prometheus.rs [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+	/// Block time measurements for relay parent blocks
+	block_times: HistogramVec,
+	/// Number of slow availability events.
+	slow_avail_count: IntCounterVec,
+	/// Number of low bitfield propagation events.
+	low_bitfields_count: IntCounterVec,
+	/// Number of bitfields being set
+	bitfields: IntGaugeVec,
+	/// Average included time in relay parent blocks
+	included_times: HistogramVec,
```
[the remark this answers is not in the document]
Comment (vstakhov) [reply src/pc/prometheus.rs]: I have renamed it and slightly adjusted buckets, as buckets for seconds and blocks should be intuitively different.

**Review by sandreim — APPROVED** [review]

**Pull request merged**